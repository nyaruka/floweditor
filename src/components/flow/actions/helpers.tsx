import {
  Contact,
  Endpoints,
  Group,
  RecipientsAction,
  FlowIssue,
  FlowIssueType,
  BroadcastMsg,
  SendMsg
} from 'flowTypes';
import * as React from 'react';
import { Asset, AssetType } from 'store/flowContext';
import { FormEntry, NodeEditorSettings, ValidationFailure } from 'store/nodeEditor';
import { createUUID } from 'utils';
import { Trans } from 'react-i18next';
import shared from 'components/shared.module.scss';
import { showHelpArticle } from 'external';
import { IssueProps } from '../props';

export const renderIssues = (issueProps: IssueProps): JSX.Element => {
  const { issues, helpArticles } = issueProps;
  if (!issues || issues.length === 0) {
    return null;
  }

  return (
    <div style={{ padding: '10px 0px' }}>
      {issues.map((issue: FlowIssue, num: Number) => {
        const key = issue.node_uuid + issue.action_uuid + num;
        return (
          <div
            style={{ margin: '6px 0px', display: 'flex', fontSize: '14px', color: 'tomato' }}
            key={key}
          >
            <div
              style={{ marginRight: '8px', marginTop: '-2px', fontSize: '18px' }}
              className={`fe-warning`}
            />
            <div>{renderIssue(issue, helpArticles)}</div>
          </div>
        );
      })}
    </div>
  );
};

export const renderIssue = (
  issue: FlowIssue,
  helpArticles: { [key: string]: string } = {}
): JSX.Element => {
  // worst case, defer to the default description
  let message: JSX.Element = <>{issue.description}</>;

  if (issue.type === FlowIssueType.MISSING_DEPENDENCY) {
    message = (
      <Trans
        i18nKey="issues.missing_dependency"
        values={{
          name: issue.dependency.name || issue.dependency.key,
          type: issue.dependency.type
        }}
      >
        Cannot find a [[type]] for <span className="emphasize">[[name]]</span>
      </Trans>
    );
  }

  if (issue.type === FlowIssueType.INVALID_REGEX) {
    message = (
      <Trans i18nKey="issues.legacy_extra" values={{ regex: issue.regex }}>
        Invalid regular expression found: [[regex]]
      </Trans>
    );
  }

  if (issue.type === FlowIssueType.LEGACY_EXTRA) {
    message = (
      <Trans i18nKey="issues.legacy_extra">Expressions should not reference @legacy_extra</Trans>
    );
  }

  const article = helpArticles[issue.type];
  if (article) {
    return (
      <div
        className={shared.issue_help}
        onClick={() => {
          showHelpArticle(article);
        }}
      >
        {message}
      </div>
    );
  }

  return message;
};

export const getActionUUID = (nodeSettings: NodeEditorSettings, currentType: string): string => {
  if (nodeSettings.originalAction && nodeSettings.originalAction.type === currentType) {
    return nodeSettings.originalAction.uuid;
  }
  return createUUID();
};

export const getCompose = (action: SendMsg | BroadcastMsg = null): string => {
  if (!action) {
    return getEmptyComposeValue();
  }
  return action.compose;
};

export const getEmptyComposeValue = (): string => {
  return JSON.stringify({ text: '', attachments: [] });
};

export const getRecipients = (action: RecipientsAction): Asset[] => {
  let selected: any[] = (action.groups || []).map((group: Group) => {
    return {
      id: group.uuid,
      name: group.name,
      type: AssetType.Group
    };
  });

  selected = selected.concat(
    (action.contacts || []).map((contact: Contact) => {
      return { id: contact.uuid, name: contact.name, type: AssetType.Contact, missing: false };
    })
  );

  selected = selected.concat(
    (action.legacy_vars || []).map((expression: string) => {
      return { name: expression, value: expression, expression: true };
    })
  );

  return selected;
};

export const renderAssetList = (
  assets: Asset[],
  max: number = 10,
  endpoints: Endpoints
): JSX.Element[] => {
  // show our missing ones first
  return assets.reduce((elements, asset, idx) => {
    if (idx <= max - 2 || assets.length === max) {
      elements.push(renderAsset(asset, endpoints));
    } else if (idx === max - 1) {
      elements.push(<div key="ellipses">+{assets.length - max + 1} more</div>);
    }
    return elements;
  }, []);
};

export const renderAsset = (asset: Asset, endpoints: Endpoints) => {
  let assetBody = null;

  switch (asset.type) {
    case AssetType.Classifier:
      assetBody = (
        <Trans i18nKey="assets.classifier" values={{ name: asset.name }}>
          Call [[name]] classifier
        </Trans>
      );
      break;
    case AssetType.Group:
      assetBody = (
        <>
          <span className={`${shared.node_group} fe-group`} />
          {asset.name}
        </>
      );
      break;
    case AssetType.Label:
      assetBody = (
        <>
          <span className={`${shared.node_label} fe-label`} />
          {asset.name}
        </>
      );
      break;
    case AssetType.Flow:
      assetBody = (
        <>
          <span className={`${shared.node_label} fe-split`} />
          <a
            onMouseDown={(e: any) => {
              e.preventDefault();
              e.stopPropagation();
            }}
            onMouseUp={(e: any) => {
              e.preventDefault();
              e.stopPropagation();
            }}
            href={`${endpoints.editor}/${asset.id}`}
            rel="noopener noreferrer"
            target="_blank"
          >
            {asset.name}
          </a>
        </>
      );
      break;
    case AssetType.Ticketer:
      assetBody = (
        <Trans i18nKey="assets.ticketer" values={{ name: asset.name }}>
          Open a new Ticket on [[name]]
        </Trans>
      );
      break;
  }

  if (!assetBody) {
    assetBody = asset.name;
  }

  return (
    <div className={`${shared.node_asset}`} key={asset.id || (asset as any).value}>
      {assetBody}
    </div>
  );
};

export const getAllErrors = (entry: FormEntry): ValidationFailure[] => {
  return entry ? entry.validationFailures || [] : [];
};

export const hasErrors = (entry: FormEntry): boolean => {
  return getAllErrors(entry).length > 0;
};

export const getAllErrorMessages = (entry: FormEntry): string[] => {
  const errors = getAllErrors(entry).map((failure: ValidationFailure) => failure.message);
  return errors;
};

export const getExpressions = (assets: any[]): any[] => {
  return assets
    .filter((asset: any) => asset.expression)
    .map((asset: any) => {
      return asset.value;
    });
};

export const getRecipientsByAsset = (assets: Asset[], type: AssetType): any[] => {
  return assets
    .filter((asset: Asset) => asset.type === type)
    .map((asset: Asset) => {
      return { uuid: asset.id, name: asset.name };
    });
};

export const getComposeByAsset = (value: string, asset: string): any | any[] => {
  return JSON.parse(value)[asset];
};
