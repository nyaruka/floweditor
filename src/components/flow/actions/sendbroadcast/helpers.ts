import {
  getActionUUID,
  getComposeActionToState,
  getComposeStateToAction,
  getRecipients,
  getRecipientsStateToAction
} from 'components/flow/actions/helpers';
import { SendBroadcastFormState } from 'components/flow/actions/sendbroadcast/SendBroadcastForm';
import { Types } from 'config/interfaces';
import { BroadcastMsg } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): SendBroadcastFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.send_broadcast) {
    let action = settings.originalAction as BroadcastMsg;

    // check if our form should use a localized action
    if (settings.localizations && settings.localizations.length > 0) {
      const localized = settings.localizations[0];
      if (localized.isLocalized()) {
        action = settings.localizations[0].getObject() as BroadcastMsg;
      } else {
        return {
          compose: { value: getComposeActionToState() },
          recipients: { value: [] },
          valid: true
        };
      }
    }

    return {
      compose: { value: getComposeActionToState(action) },
      recipients: { value: getRecipients(action) },
      valid: true
    };
  }

  return {
    compose: { value: getComposeActionToState() },
    recipients: { value: [] },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: SendBroadcastFormState
): BroadcastMsg => {
  const [compose, text, attachments] = getComposeStateToAction(formState);
  const [legacy_vars, contacts, groups] = getRecipientsStateToAction(formState);

  return {
    legacy_vars: legacy_vars,
    contacts: contacts,
    groups: groups,
    compose: compose,
    text: text,
    attachments: attachments,
    type: Types.send_broadcast,
    uuid: getActionUUID(settings, Types.send_broadcast)
  };
};
