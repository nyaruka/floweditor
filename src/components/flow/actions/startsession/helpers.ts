import {
  getActionUUID,
  getExpressions,
  getRecipients,
  getRecipientsByAsset
} from 'components/flow/actions/helpers';
import {
  StartSessionFormState,
  START_TYPE_CREATE,
  START_TYPE_ASSETS,
  START_TYPE_QUERY
} from 'components/flow/actions/startsession/StartSessionForm';
import { Types } from 'config/interfaces';
import { StartSession } from 'flowTypes';
import { AssetType } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): StartSessionFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.start_session) {
    const action = settings.originalAction as StartSession;

    const init = {
      recipients: {
        value: getRecipients(action)
      },
      flow: {
        value: action.flow
      },
      startType: {
        value: action.create_contact
          ? START_TYPE_CREATE
          : action.contact_query
          ? START_TYPE_QUERY
          : START_TYPE_ASSETS
      },
      contactQuery: { value: action.contact_query || '' },
      exclusions: action.exclusions,
      valid: true
    };

    return init;
  }

  return {
    recipients: { value: [] },
    flow: { value: null },
    startType: { value: START_TYPE_ASSETS },
    contactQuery: { value: '' },
    exclusions: { in_a_flow: false },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  state: StartSessionFormState
): StartSession => {
  const flow = state.flow.value;

  const action: StartSession = {
    legacy_vars: getExpressions(state.recipients.value),
    contacts: getRecipientsByAsset(state.recipients.value, AssetType.Contact),
    groups: getRecipientsByAsset(state.recipients.value, AssetType.Group),
    create_contact: state.startType.value === START_TYPE_CREATE,
    flow: { name: flow.name, uuid: flow.uuid },
    type: Types.start_session,
    uuid: getActionUUID(settings, Types.start_session),
    exclusions: state.exclusions
  };

  // only include contact query if it is set
  if (state.contactQuery.value) {
    action['contact_query'] = state.contactQuery.value;
  }

  return action;
};
