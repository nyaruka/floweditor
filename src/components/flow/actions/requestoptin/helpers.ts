import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { RequestOptIn } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';
import { RequestOptInFormState } from './RequestOptInForm';

export const initializeForm = (settings: NodeEditorSettings): RequestOptInFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.request_optin) {
    const action = settings.originalAction as RequestOptIn;
    return {
      optin: { value: action.optin },
      valid: true
    };
  }

  return {
    optin: { value: null },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: RequestOptInFormState
): RequestOptIn => {
  const result = {
    type: Types.request_optin,
    optin: formState.optin.value[0],
    uuid: getActionUUID(settings, Types.add_input_labels)
  };
  return result;
};
