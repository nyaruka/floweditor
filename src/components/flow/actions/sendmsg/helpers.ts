/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { getActionUUID } from 'components/flow/actions/helpers';
import { SendMsgFormState } from 'components/flow/actions/sendmsg/SendMsgForm';
import { Types } from 'config/interfaces';
import { FlowEditorConfig, Label, MsgTemplating, SendMsg } from 'flowTypes';
import { FormEntry, NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { SelectOption } from 'components/form/select/SelectElement';
import { createUUID } from 'utils';
import { Attachment } from './attachments';

export const TOPIC_OPTIONS: SelectOption[] = [
  { value: 'event', name: 'Event' },
  { value: 'account', name: 'Account' },
  { value: 'purchase', name: 'Purchase' },
  { value: 'agent', name: 'Agent' }
];

export const initializeForm = (
  settings: NodeEditorSettings,
  config: FlowEditorConfig
): SendMsgFormState => {
  let template: FormEntry = { value: null };
  let templateVariables: StringEntry[] = [];
  let skipValidation = false;

  if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
    const action = settings.originalAction as SendMsg;
    const attachments: Attachment[] = [];
    let expressionValue = null;
    (action.attachments || []).forEach((attachmentString: string) => {
      const splitPoint = attachmentString.indexOf(':');

      const type = attachmentString.substring(0, splitPoint);
      const attachment = {
        type,
        url: attachmentString.substring(splitPoint + 1),
        uploaded: type.indexOf('/') > -1,
        valid: false
      };

      attachments.push(attachment);
    });

    if (action.templating) {
      const msgTemplate = action.templating.template;
      if (action.templating.expression) {
        expressionValue = { value: action.templating.expression };
      }
      template = {
        value: {
          uuid: msgTemplate.uuid,
          name: msgTemplate.name
        }
      };
      templateVariables = action.templating.variables.map((value: string) => {
        return {
          value
        };
      });
    }

    const labels = action.labels
      ? action.labels.map((label: Label) => {
          if (label.name_match) {
            return { name: label.name_match, expression: true };
          }
          return label;
        })
      : [];

    if (config.skipValidation) {
      skipValidation = config.skipValidation;
    }

    return {
      expression: expressionValue,
      topic: { value: TOPIC_OPTIONS.find(option => option.value === action.topic) },
      template,
      templateVariables,
      attachments,
      labels: {
        value: labels
      },
      uploadInProgress: false,
      uploadError: '',
      message: { value: action.text },
      quickReplies: { value: action.quick_replies || [] },
      quickReplyEntry: { value: '' },
      sendAll: action.all_urns,
      valid: true,
      skipValidation
    };
  }

  return {
    topic: { value: null },
    template,
    templateVariables: [],
    attachments: [],
    uploadInProgress: false,
    uploadError: '',
    message: { value: '' },
    quickReplies: { value: [] },
    quickReplyEntry: { value: '' },
    sendAll: false,
    valid: false,
    labels: { value: [] }
  };
};

export const stateToAction = (settings: NodeEditorSettings, state: SendMsgFormState): SendMsg => {
  const attachments = state.attachments
    .filter((attachment: Attachment) => attachment.url.trim().length > 0)
    .map((attachment: Attachment) => `${attachment.type}:${attachment.url}`);

  let templating: MsgTemplating = null;

  if (state.template && state.template.value) {
    let templatingUUID = createUUID();
    if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
      const action = settings.originalAction as SendMsg;
      if (
        action.templating &&
        action.templating.template &&
        action.templating.template.uuid === state.template.value.id
      ) {
        templatingUUID = action.templating.uuid;
      }
    }

    templating = {
      uuid: templatingUUID,
      template: {
        uuid: state.template.value.uuid,
        name: state.template.value.name
      },
      variables: state.templateVariables.map((variable: StringEntry) => variable.value)
    };

    if (state.expression && state.expression.value) {
      templating.expression = state.expression.value;
    }
  }
  const result: SendMsg = {
    attachments,
    text: state.message.value,
    type: Types.send_msg,
    all_urns: state.sendAll,
    labels: state.labels.value.map((label: any) => {
      if (label.expression) {
        return { name_match: label.name };
      }
      return label;
    }),
    quick_replies: state.quickReplies.value,
    uuid: getActionUUID(settings, Types.send_msg)
  };

  if (templating) {
    result.templating = templating;
  }

  if (state.topic.value) {
    result.topic = state.topic.value.value;
  }

  return result;
};
