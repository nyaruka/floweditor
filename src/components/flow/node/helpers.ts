import { getSwitchRouter } from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import {
  CallResthook,
  CallWebhook,
  Category,
  Exit,
  FlowNode,
  RouterTypes,
  TransferAirtime,
  Action,
  AnyAction,
  FlowIssue
} from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { getType } from 'config/typeConfigs';

export const getCategoriesForExit = (renderNode: RenderNode, exit: Exit): Category[] => {
  // if we are
  const isGroupSplit = getType(renderNode) === Types.split_by_groups;

  if (!renderNode.node.router) {
    return [];
  }
  return renderNode.node.router.categories
    .filter((cat: Category) => cat.exit_uuid === exit.uuid)
    .map((cat: Category) => {
      if (isGroupSplit) {
        return {
          ...cat
        };
      } else {
        return { ...cat, missing: false };
      }
    });
};

export const getResultName = (node: FlowNode) => {
  const switchRouter = getSwitchRouter(node);

  if (switchRouter && switchRouter.result_name) {
    return switchRouter.result_name;
  }

  if (node.router && node.router.type === RouterTypes.random) {
    return node.router.result_name;
  }

  if (node.actions.length === 1) {
    const action = node.actions[0];
    if (
      action.type === Types.call_webhook ||
      action.type === Types.call_resthook ||
      action.type === Types.open_ticket ||
      action.type === Types.transfer_airtime ||
      action.type === Types.link_google_sheet
    ) {
      const resultAction = action as CallWebhook | CallResthook | TransferAirtime;
      return resultAction.result_name;
    }
  }
};

export const getVisibleActions = (renderNode: RenderNode): Action[] => {
  // subflow nodes hide their set run results
  if (getType(renderNode) === Types.split_by_subflow) {
    return renderNode.node.actions.filter((action: Action) => action.type !== Types.set_run_result);
  }

  return renderNode.node.actions;
};

export const filterIssuesForAction = (
  nodeUUID: string,
  action: AnyAction,
  issues: FlowIssue[]
): FlowIssue[] => {
  return issues.filter(issue => issue.node_uuid === nodeUUID && issue.action_uuid === action.uuid);
};
