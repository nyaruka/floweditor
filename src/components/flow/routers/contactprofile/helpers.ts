import { getActionUUID } from 'components/flow/actions/helpers';
import { createWebhookBasedNode } from 'components/flow/routers/helpers';

import { Types } from 'config/interfaces';
import { SetContactProfile } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

import { ContactProfileRouterFormState, profileOptions } from './ContactProfileRouterForm';

export const nodeToState = (settings: NodeEditorSettings): ContactProfileRouterFormState => {
  let resulNode: ContactProfileRouterFormState = {
    valid: true,
    optionType: { value: profileOptions['1'] },
    profileName: { value: '' },
    profileType: { value: '' }
  };

  if (settings.originalAction && settings.originalAction.type === Types.set_contact_profile) {
    const action = settings.originalAction as SetContactProfile;

    if (action.profile_type) {
      const option =
        Object.values(profileOptions).find(value => value.name === action.profile_type) ||
        profileOptions['1'];

      resulNode.optionType = {
        value: option
      };

      if (typeof action.value === 'string') {
        resulNode.profileName = { value: action.value };
      } else {
        resulNode.profileName = { value: action.value.name };
        resulNode.profileType = { value: action.value.type };
      }
    }
  }

  return resulNode;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ContactProfileRouterFormState
): RenderNode => {
  const { optionType, profileName, profileType } = state;

  const newAction: SetContactProfile = {
    profile_type: optionType.value.name,
    result_name: profileName.value,
    value:
      optionType.value.name === 'Create Profile'
        ? { name: profileName.value, type: profileType.value }
        : profileName.value,
    type: Types.set_contact_profile,
    uuid: getActionUUID(settings, Types.set_contact_profile)
  };

  return createWebhookBasedNode(newAction, settings.originalNode, false);
};
