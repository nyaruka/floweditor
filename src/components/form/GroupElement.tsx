import * as React from 'react';
import { v4 as generateUUID } from 'uuid';
import Select from 'react-select';
import { SearchResult } from '../../services/ComponentMap';
import FormElement, { FormElementProps } from './FormElement';
import SelectSearch from '../SelectSearch';
import { getSelectClass } from '../../helpers/utils';

import * as styles from './FormElement.scss';

interface GroupOption {
    group: string;
    name: string;
}

export type GroupList = GroupOption[];

interface GroupElementProps extends FormElementProps {
    groups: GroupList;
    localGroups?: SearchResult[];
    endpoint?: string;
    add?: boolean;
    placeholder?: string;
    searchPromptText?: string;
}

interface GroupElementState {
    groups: SearchResult[];
    errors: string[];
}

export const transformGroups = (groups: GroupList): SearchResult[] =>
    groups.map(({ name, group }) => ({ name, id: group, type: 'group' }));

export default class GroupElement extends React.Component<GroupElementProps, GroupElementState> {
    constructor(props: GroupElementProps) {
        super(props);

        const groups = transformGroups(this.props.groups);

        this.state = {
            groups,
            errors: []
        };

        this.onChange = this.onChange.bind(this);
        this.isValidNewOption = this.isValidNewOption.bind(this);
        this.createNewOption = this.createNewOption.bind(this);
    }

    private onChange(groups: SearchResult[]): void {
        this.setState({
            groups
        });
    }

    public validate(): boolean {
        const errors: string[] = [];
        const { groups } = this.state;

        if (this.props.required) {
            if (groups.length === 0) {
                errors.push(`${this.props.name} is required`);
            }
        }

        this.setState({ errors });

        return errors.length === 0;
    }

    private isValidNewOption({ label }: { label: string }): boolean {
        if (!label) {
            return false;
        }

        const lowered = label.toLowerCase();

        return lowered.length > 0 && lowered.length <= 36 && /^[a-z0-9-][a-z0-9- ]*$/.test(lowered);
    }

    private createNewOption(arg: { label: string }): SearchResult {
        const newOption = {
            id: generateUUID(),
            name: arg.label,
            extraResult: true
        } as SearchResult;

        return newOption;
    }

    public render(): JSX.Element {
        const createOptions: any = {};

        if (this.props.add) {
            createOptions.isValidNewOption = this.isValidNewOption;
            createOptions.createNewOption = this.createNewOption;
            createOptions.createPrompt = 'New group: ';
        }

        const className: string = getSelectClass(this.state.errors.length);

        return (
            <FormElement name={this.props.name} errors={this.state.errors}>
                <SelectSearch
                    className={className}
                    onChange={this.onChange}
                    name={this.props.name}
                    url={this.props.endpoint}
                    resultType="group"
                    localSearchOptions={this.props.localGroups}
                    multi={true}
                    initial={this.state.groups}
                    closeOnSelect={false}
                    placeholder={this.props.placeholder}
                    searchPromptText={this.props.searchPromptText}
                    {...createOptions}
                />
            </FormElement>
        );
    }
}
