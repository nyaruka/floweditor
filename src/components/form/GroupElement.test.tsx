import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import GroupElement, {
    GroupElementProps,
    GroupList,
    transformGroups,
    isValidNewOption,
    createNewOption
} from './GroupElement';
import Config from '../../providers/ConfigProvider/configContext';
import { GROUP_PLACEHOLDER, GROUP_NOT_FOUND } from '../routers/SwitchRouter';
import { SearchResult } from '../../services/ComponentMap';
import { validUUID } from '../../helpers/utils';
import { transformGroups } from './GroupElement';

const { results: groupsResp } = require('../../../assets/groups.json');

describe('GroupElement >', () => {
    const props: GroupElementProps = {
        name: 'Group',
        endpoint: Config.endpoints,
        placeholder: GROUP_PLACEHOLDER,
        searchPromptText: GROUP_NOT_FOUND,
        required: true
    };

    const groupList: GroupList = groupsResp.map(({ name, uuid }) => ({ name, group: uuid }));

    describe('helpers >', () => {
        const newGroup: { label: string } = { label: 'new group' };

        describe('transformGroups >', () => {
            it('should transform group list into search result objects', () => {
                expect(transformGroups(groupList)).toMatchSnapshot();
            });
        });

        describe('isValidNewOption >', () => {
            it('should return false if new option is invalid', () => {
                expect(isValidNewOption({ label: '$$$' })).toBeFalsy();
            });

            it('should return true if new option is valid', () => {
                expect(isValidNewOption(newGroup)).toBeTruthy();
            });
        });

        describe('createNewOption >', () => {
            it('should generate a new search result object', () => {
                const newOption: SearchResult = createNewOption(newGroup);

                expect(validUUID(newOption.id)).toBeTruthy();
                expect(newOption.name).toBe(newGroup.label);
                expect(newOption.extraResult).toBeTruthy();
            });
        });
    });

    describe('render >', () => {
        describe('split by group membership', () => {
            it('should render <SelectSearch />, pass it expected props', () => {
                const wrapper: ReactWrapper = mount(<GroupElement {...props} />);

                expect(wrapper.find('SelectSearch').props()).toEqual(
                    expect.objectContaining({
                        className: '',
                        onChange: wrapper.instance().onChange,
                        name: props.name,
                        url: props.endpoint,
                        resultType: 'group',
                        multi: true,
                        initial: [],
                        closeOnSelect: false,
                        placeholder: props.placeholder,
                        searchPromptText: props.searchPromptText
                    })
                );
            });
        });
    });

    describe('instance methods >', () => {
        const groups: SearchResult[] = transformGroups(groupList).slice(3);
        const searchResults: SearchResult[] = transformGroups(groupList);

        describe('onChange >', () => {
            it('should update state when called', () => {
                const wrapper: ReactWrapper = mount(<GroupElement {...props} />);

                wrapper.instance().onChange(groups);

                expect(wrapper.state('groups')).toEqual(groups);
            });

            it("should call 'onChange' prop if passed", () => {
                const onChange: jest.Mock<{}> = jest.fn();
                const wrapper: ReactWrapper = mount(<GroupElement {...{ ...props, onChange }} />);

                wrapper.instance().onChange(groups);

                expect(onChange).toHaveBeenCalled();
            });
        });

        describe('validate >', () => {
            it("should return false, update state if GroupElement isn't valid", () => {
                const wrapper: ReactWrapper = mount(<GroupElement {...props} />);

                expect(wrapper.instance().validate()).toBeFalsy();
                expect(wrapper.state('errors')[0]).toBe('Group is required');
            });

            it('should return true if GroupElement is valid', () => {
                const wrapper: ReactWrapper = mount(<GroupElement {...props} />);

                expect(wrapper.instance().onChange(groups));
                expect(wrapper.instance().validate()).toBeTruthy();
            });
        });

        describe('getGroups >', () => {
            it("should return an empty array if GroupElement is passed a falsy 'groups' prop", () => {
                const wrapper: ReactWrapper = mount(<GroupElement {...props} />);

                expect(wrapper.instance().getGroups()).toEqual([]);
            });

            it("should return an array of SearchResult objects if GroupElement is passed a truthy 'groups' prop", () => {
                const wrapper: ReactWrapper = mount(
                    <GroupElement {...{ ...props, groups: groupList }} />
                );

                expect(wrapper.instance().getGroups()).toEqual(searchResults);
            });

            it("should return localGroups array if GroupElement passed 'localGroups' prop but not 'groups'", () => {
                const wrapper: ReactWrapper = mount(
                    <GroupElement {...{ ...props, localGroups: searchResults }} />
                );

                expect(wrapper.instance().getGroups()).toEqual(searchResults);
            });
        });
    });
});
