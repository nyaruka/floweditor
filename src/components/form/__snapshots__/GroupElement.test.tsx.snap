// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupElement > helpers > transformGroups > should transform group list into search result objects 1`] = `
Array [
  Object {
    "id": "23ff7152-b588-43e4-90de-fda77aeaf7c0",
    "name": "Customers",
    "type": "group",
  },
  Object {
    "id": "2429d573-80d7-47f8-879f-f2ba442a1bfd",
    "name": "Unsatisfied Customers",
    "type": "group",
  },
  Object {
    "id": "cdbf9e01-aaa7-4381-8259-ee042447bcac",
    "name": "Early Adopters",
    "type": "group",
  },
  Object {
    "id": "afaba971-8943-4dd8-860b-3561ed4f1fe1",
    "name": "Testers",
    "type": "group",
  },
  Object {
    "id": "33b28bac-b588-43e4-90de-fda77aeaf7c0",
    "name": "Subscribers",
    "type": "group",
  },
]
`;
